src/editor/codemirror/edits.ts

```diff
@@ -77,59 +73,93 @@ export const calculateChanges = (state: EditorState, addition: string) => {
   );
   const allCurrentImports = currentImports(state);
 
-  const insertPoint = defaultImportInsertPoint(state, allCurrentImports);
+  const importInsertPoint = defaultInsertPoint(state, allCurrentImports);
   const changes = requiredImports.flatMap((required) =>
-    calculateImportChangesInternal(insertPoint, allCurrentImports, required)
+    calculateImportChangesInternal(
+      importInsertPoint,
+      allCurrentImports,
+      required
+    )
   );
-  if (insertPoint.used) {
-    changes[0].insert = insertPoint.prefix + changes[0].insert;
-  }
   if (changes.length > 0 && allCurrentImports.length === 0) {
     // Two blank lines separating the imports from everything else.
     changes[changes.length - 1].insert += "\n\n";
   }
+  let importLines = changes
+    .map((c) => c.insert.split("\n").length - 1)
+    .reduce((acc, cur) => acc + cur, 0);
 
-  // We'll want to add more sophisticated insertion than this.
+  let additionInsertPoint: number = -1;
   if (addition) {
-    // We always want a newline at the end as for now our inserts are whole lines.
-    // If the insertion point is after a node then we want an newline before it.
-    // If the insertion point is before a node then we want an newline after it to break the line.
-    const lastImport = allCurrentImports?.[allCurrentImports.length - 1]?.node;
-    let relation = Relation.Default;
-    let insertionPoint = 0;
-    if (lastImport?.nextSibling) {
-      relation = Relation.Before;
-      insertionPoint = lastImport.nextSibling.from;
-    } else if (lastImport) {
-      relation = Relation.After;
-      insertionPoint = lastImport.to;
+    let additionPrefix = "";
+    if (line !== undefined) {
+      // Tweak so the addition preview is under the mouse even if we added imports.
+      line -= importLines;
+      line = Math.max(1, line);
+      const extraLines = line - state.doc.lines;
+      if (extraLines > 0) {
+        additionInsertPoint = state.doc.length;
+        additionPrefix = "\n".repeat(extraLines);
+      } else {
+        additionInsertPoint = state.doc.line(line).from;
+      }
+    } else {
+      additionInsertPoint = skipWhitespaceLines(
```

Comment:
I can't really remember why we do this. Remove it and see what fails 😄

Reply:
This is covered by the "non-import content before code rather than after imports" case. I've added an explanatory comment.